### PR TITLE
refactor(plugin-js-packages): use semver for version diff

### PR DESCRIPTION
**Pull request by Tlacenka**
Related to #597 

In this PR, I use [`semver` functions](https://github.com/npm/node-semver?tab=readme-ov-file#functions) to determine the version difference.

### CODE REVIEW

File: packages/plugin-js-packages/src/lib/runner/outdated/transform.ts [hunk 1]
```diff
@@ -18,23 +14,27 @@ export function outdatedResultToAuditOutput(
...
-    );
 
-  const outdatedStats = outdatedDependencies.reduce(
-    (acc, dep) => {
-      const outdatedLevel = getOutdatedLevel(dep.current, dep.latest);
-      return { ...acc, [outdatedLevel]: acc[outdatedLevel] + 1 };
-    },
-    { major: 0, minor: 0, patch: 0 },
+  const validDependencies = relevantDependencies
+    .filter(dep => parse(dep.current) != null && parse(dep.latest) != null)
```
Comment (BioPhoton) [inline packages/plugin-js-packages/src/lib/runner/outdated/transform.ts:19]: Could you comment what you filter out pls

**Review by BioPhoton — DISMISSED** [review]

**Comment by BioPhoton** [comment]
> Can I get rid of the non-null-assertion? I use semver functions to make sure to filter out null values and up-to-date dependencies but TS does not know.

Where? Maybe a guard helps?

File: packages/plugin-js-packages/src/lib/runner/outdated/transform.ts [hunk 2]
```diff
@@ -18,23 +14,27 @@ export function outdatedResultToAuditOutput(
...
+    .map(dep => ({
+      ...dep,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      current: clean(dep.current)!,
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      latest: clean(dep.latest)!,
+    }));
```
Comment (matejchalk) [inline packages/plugin-js-packages/src/lib/runner/outdated/transform.ts:26]: > Can I get rid of the non-null-assertion? I use semver functions to make sure to filter out null values and up-to-date dependencies but TS does not know.

You can use a type predicate if you swap the order - first call `clean()` in `.map()`, then check both are non-null (i.e. valid) in `.filter()`. According to the docs, both `parse()` and `clean()` return `null` for invalid versions, so don't need to call both.

```suggestion
  const validDependencies = relevantDependencies
    .map(dep => ({
      ...dep,
      current: clean(dep.current),
      latest: clean(dep.latest),
    }))
    .filter(
      (dep): dep is OutdatedResult[number] =>
        dep.current != null && dep.latest != null,
    );
```

File: packages/plugin-js-packages/src/lib/runner/outdated/transform.ts [hunk 3]
```diff
@@ -18,23 +14,27 @@ export function outdatedResultToAuditOutput(
...
+
+  const outdatedDependencies = validDependencies.filter(dep =>
+    neq(dep.current, dep.latest),
   );
 
+  const outdatedStats = outdatedDependencies.reduce((acc, dep) => {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const outdatedLevel = diff(dep.current, dep.latest)!;
+    return { ...acc, [outdatedLevel]: acc[outdatedLevel] + 1 };
+  }, objectFromEntries(RELEASE_TYPES.map(versionType => [versionType, 0])));
```
Comment (matejchalk) [inline packages/plugin-js-packages/src/lib/runner/outdated/transform.ts:36]: The docs say that `diff()` returns `null` if the versions are the same. That's probably not gonna happen in `npm outdated`'s output. But to be on the safe-side, such `diff()` results can easily be ignored:

```suggestion
  const outdatedStats = outdatedDependencies.reduce((acc, dep) => {
    const outdatedLevel = diff(dep.current, dep.latest);
    if (outdatedLevel == null) {
      return acc;
    }
    return { ...acc, [outdatedLevel]: acc[outdatedLevel] + 1 };
  }, objectFromEntries(RELEASE_TYPES.map(versionType => [versionType, 0])));
```

File: packages/plugin-js-packages/src/lib/runner/outdated/constants.ts [hunk 4]
```diff
@@ -1,8 +1,12 @@
+import { ReleaseType } from 'semver';
 import type { IssueSeverity } from '@code-pushup/models';
-import { VersionType } from './types';
 
-export const outdatedSeverity: Record<VersionType, IssueSeverity> = {
+export const outdatedSeverity: Record<ReleaseType, IssueSeverity> = {
   major: 'error',
+  premajor: 'error', // TODO ??
   minor: 'warning',
+  preminor: 'warning',
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply packages/plugin-js-packages/src/lib/runner/outdated/constants.ts:8]: Makes sense.

File: packages/plugin-js-packages/src/lib/runner/outdated/transform.ts (hunk 2 above)
Comment (Tlacenka) [reply packages/plugin-js-packages/src/lib/runner/outdated/transform.ts:26]: Nice, thank you

File: packages/plugin-js-packages/src/lib/runner/outdated/transform.ts (hunk 1 above)
Comment (Tlacenka) [reply packages/plugin-js-packages/src/lib/runner/outdated/transform.ts]: I updated the code.

File: packages/plugin-js-packages/src/lib/runner/outdated/transform.ts [hunk 5]
```diff
@@ -43,7 +43,7 @@ export function outdatedResultToAuditOutput(
   return {
     slug: `${packageManager}-outdated-${depGroup}`,
     score: calculateOutdatedScore(
-      outdatedStats.major,
+      outdatedStats.major + outdatedStats.premajor,
```
[the remark this answers is not in the document]
Comment (Tlacenka) [reply packages/plugin-js-packages/src/lib/runner/outdated/transform.ts]: Addressed.

**Review by matejchalk — APPROVED** [review]
Good work, looks very robust :+1:

**Pull request merged**